Add support for NFS server

darkowlzz opened the pull request:
Adds a new custom resource `nfsservers.storageos.com` for managing NFS server resources.

A StorageOS Cluster resource is a prerequisite for NFS server because the underlying storage requires StorageOS volumes.
On creating a NFSServer resource, the NFS server controller provisions NFS ganesha server instance with a StorageOS volume attached.

Short name for `nfsservers.storageos.com` is `nfsserver`.

Status of provisioned NFS server:
```
$ kubectl get nfsserver
NAME                STATUS    CAPACITY   TARGET         ACCESS MODES   STORAGECLASS   AGE
example-nfsserver   Running   1Gi        10.103.56.29   RW             fast           52m
```

croomes commented on pkg/nfs/configmap.go:
```diff
@@ -0,0 +1,208 @@
+		id++
+	}
+
+	globalCfg, err := globalConfig(true, true)
```
Use consts for graceless and fsidDevice?

croomes commented on pkg/nfs/configmap.go:
```diff
@@ -0,0 +1,208 @@
+		}
+		exports = append(exports, exportCfg)
+		id++
+	}
```
I think we can simplify this a bit as we've settled on having a single export per pv.  I don't think we will need to support multiple exports per server.

darkowlzz replied:
Will remove this in a separate PR.

darkowlzz replied:
Remove multiple exports support and changed it to customizable single export in this PR itself.

croomes commented on pkg/nfs/deploy.go:
```diff
@@ -0,0 +1,63 @@
+	statefulsetKind = "statefulset"
+
+	DefaultNFSPort     = 2049
+	DefaultRPCPort     = 111
```
we can remove `DefaultRPCPort`.  Only 2049 and 9587 are needed.

croomes commented on pkg/nfs/statefulset.go:
```diff
@@ -0,0 +1,167 @@
+						{
+							Name:          "rpc-port",
+							ContainerPort: int32(rpcPort),
+						},
```
can remove rpc-port

croomes commented on pkg/nfs/statefulset.go:
```diff
@@ -0,0 +1,167 @@
+					ImagePullPolicy: "IfNotPresent",
+					Name:            "nfsd",
+					Image:           d.nfsServer.Spec.GetContainerImage(),
+					// Args: []string{"nfs", "server", "--ganeshaConfigPath=" + NFSConfigMapPath + "/" + nfsServer.name},
```
Args can be removes - new container will only use env vars

croomes commented on pkg/nfs/statefulset.go:
```diff
@@ -0,0 +1,167 @@
+							Name:  "GANESHA_CONFIGFILE",
+							Value: "/config/" + d.nfsServer.Name,
+						},
+					},
```
Can you also add env vars for `NAME` and `NAMESPACE` set to the RWX PVC name and namespace?  This will be used for decorating the metrics.

darkowlzz replied:
Exact PVC name won't be accurate. Because the volume claim is a template in StatefulSet, only the PVC name prefix is predictable. PVC name is appended by the StatefulSet name and a number for the replicas.
I guess the prefix name should be enough?

croomes replied:
It should be the name of the RWX vol, not the underlying StorageOS vol, so I think you can use `d.nfsServer.Name` and `d.nfsServer.Namespace` here.

darkowlzz replied:
Added the env vars in this PR itself.

croomes commented on pkg/apis/storageos/v1/nfsserver_types.go:
```diff
@@ -0,0 +1,208 @@
+
+	// Replicas of the NFS daemon
+	// TODO(simon): Don't think we can have multiple servers?
+	// Replicas int `json:"replicas,omitempty"`
```
can remove Replicas

croomes commented on pkg/apis/storageos/v1/nfsserver_types.go:
```diff
@@ -0,0 +1,208 @@
+	Squash string `json:"squash,omitempty"`
+
+	// The clients allowed to access the NFS export
+	AllowedClients []AllowedClientsSpec `json:"allowedClients,omitempty"`
```
We should either remove `AllowedClients` or figure out how it might be used and test it.  I think it's probably best to remove it for now.

darkowlzz replied:
Keeping it for now. Will remove in a follow-up PR.

darkowlzz replied:
Removed it in this PR.

croomes commented on pkg/apis/storageos/v1/nfsserver_types.go:
```diff
@@ -0,0 +1,208 @@
+const (
+	// DefaultNFSContainerImage is the name of the Ganesha container to run.
+	// TODO: change to an image we maintain.
+	DefaultNFSContainerImage = "darkowlzz/nfs-ganesha:v0.0.2"
```
`storageos/nfs:test` is now pushed for testing.

darkowlzz commented:
CI was failing due to pod toleration.
```
  Warning  FailedScheduling  15s (x4 over 19s)  default-scheduler  0/1 nodes are available: 1 node(s) had taints that the pod didn't tolerate.
```
We taint the only node in order to test pod toleration.
Need to add pod toleration configuration for NFS server as well.

darkowlzz commented:
Added a temporary check in e2e tests as a workaround for the limited CSI support in openshift 3.11, which is required for NFS server. CSI volumes mount fail and the NFS server pod remains in pending state. SCC permissions are added. Should work on OpenShift 4.
Addressed the review suggestions.
